ipa-core/src/protocol/step/compact.rs

```diff
@@ -1,42 +1,80 @@
 use std::fmt::{Debug, Display, Formatter};
 
-use ipa_macros::Gate;
+use hex::FromHex;
 
 use super::StepNarrow;
-use crate::helpers::{prss_protocol::PrssExchangeStep, query::QueryType};
+use crate::{helpers::{prss_protocol::PrssExchangeStep, query::QueryType}, protocol::step::{GateId, GateIdArray, Step}};
 
-#[derive(Gate, Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Default)]
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Default)]
 #[cfg_attr(
     feature = "enable-serde",
     derive(serde::Deserialize),
     serde(from = "&str")
 )]
-pub struct Compact(pub u16);
+pub struct Compact {
+    id: GateId,
+}
+
+impl Compact {
+    pub fn id(&self) -> &GateId {
+        &self.id
+    }
+}
 
 // serde::Deserialize requires From<&str> implementation
 impl From<&str> for Compact {
     fn from(id: &str) -> Self {
-        // id begins with '/'. strip it.
-        Compact::deserialize(&id[1..])
+        Compact {
+            id: GateIdArray::from_hex(id.strip_prefix('/').unwrap_or(id)).unwrap().into(), // TODO use a deser implementation so we can return errors
+        }
     }
 }
 
 impl Display for Compact {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self.as_ref())
+        f.write_str(hex::encode(&self.id).as_str())
```

Comment:
is it possible to go from compact to descriptive in this implementation? compact identifiers do not carry a lot of information on their own. IIRC gates are used in HTTP layer, using base64 over hex is probably going to be hard/cryptic especially when we start working on MPC spec